PL2/PL2-shared/src/main/java/nl/tudelft/pl2016gr2/model/HashGraph.java

```diff
@@ -0,0 +1,154 @@
+package nl.tudelft.pl2016gr2.model;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map;
+
+public class HashGraph implements SequenceGraph {
+
+  private final HashMap<Integer, GraphNode> nodes;
+  private final HashSet<Integer> rootNodes;
+  private final HashSet<String> genomes;
+
+  /**
+   * Constructs an empty <code>HashGraph</code>.
+   */
+  public HashGraph() {
+    nodes = new HashMap<>();
+    rootNodes = new HashSet<>();
+    genomes = new HashSet<>();
+  }
+
+  /**
+   * Constructs a <code>HashGraph</code> with the specified elements and genomes.
+   *
+   * @param nodes   The nodes of the graph
+   * @param genomes The genomes that are represented in the graph
+   */
+  public HashGraph(Map<Integer, ? extends GraphNode> nodes, Collection<String> genomes) {
+    this.nodes = new HashMap<>(nodes);
+    this.genomes = new HashSet<>(genomes);
+    this.rootNodes = parseRootNodes();
+  }
+
+  /**
+   * Construct an <code>HashGraph</code> with the specified elements and genomes, as well as
+   * pre-determined root nodes.
+   * <p>
+   * A root node is a node that has no in-edges.
+   * </p>
+   *
+   * @param nodes     The nodes of the graph
+   * @param rootNodes The root nodes of the graph
+   * @param genomes   The genomes that are represented in the graph
+   */
+  public HashGraph(Map<Integer, ? extends GraphNode> nodes, Collection<Integer> rootNodes,
+      Collection<String> genomes) {
+    this.nodes = new HashMap<>(nodes);
+    this.rootNodes = new HashSet<>(rootNodes);
+    this.genomes = new HashSet<>(genomes);
+  }
+
+  /**
+   * Finds all root nodes in the graph.
+   * <p>
+   * This operation runs in <code>O(n)</code> for <code>n</code> nodes.
+   * </p>
+   *
+   * @return all root nodes.
+   */
+  private HashSet<Integer> parseRootNodes() {
+    Collection<Integer> rootNodes = new ArrayList<>();
+    nodes.forEach((Integer elemId, GraphNode elem) -> {
+      if (elem.isRoot()) {
+        rootNodes.add(elemId);
+      }
+    });
+    return rootNodes;
```

Comment:
I edited the `rootNodes` list to be an `ArrayList<>` rather than a `HashSet<>`. The helper function returns a new `ArrayList<>` which can directly be used as rootNode list. The `rootNodes` field is of type `ArrayList<>`, because it is pointless to hide the concrete  implementation from the class itself (by storing it as `Collection<>`). The `public` methods do return `Collection<>`s though.